### PR TITLE
[feature]: add embedding parallel

### PR DESCRIPTION
**Pull request by chengmengli06**
1. add support for dense embedding parallel
2. add support for sparse(sok) embedding parallel

### CODE REVIEW

File: easy_rec/python/utils/input_utils.py [hunk 1]
```diff
@@ -72,3 +73,30 @@ def string_to_number(field, ftype, default_value, name=''):
...
+  tf_types = []
+  for col in cols:
+    obj = input_data[col][0]
+    if isinstance(obj, list):
+      data_type = type(obj[0])
+    elif isinstance(obj, np.ndarray):
+      data_type = type(obj[0])
+    else:
+      data_type = type(obj)
+    if data_type in (int, np.int32):
```
Comment (poson) [inline easy_rec/python/utils/input_utils.py:90]: import tensorflow as tf

import numpy as np
dict_ = { int:tf.int32,np.int32:tf.int32,np.int64:tf.int64}
>>> dict_[np.int64]
tf.int64
用dict 查表要简洁一点

File: easy_rec/python/utils/test_utils.py [hunk 2]
```diff
@@ -600,6 +601,7 @@ def _ps_worker_distribute_eval(pipeline_config_path,
   procs = {}
   tf_config['task'] = {'type': chief_or_master, 'index': 0}
   os.environ['TF_CONFIG'] = json.dumps(tf_config)
+  os.environ[constant.SORT_COL_BY_NAME] = '1'
```
Comment (poson) [inline easy_rec/python/utils/test_utils.py:604]: '1' 是什么含义？第一个字段吗

File: easy_rec/python/utils/estimator_utils.py [hunk 3]
```diff
@@ -1009,6 +1018,16 @@ def init_hvd():
   os.environ['HOROVOD_RANK'] = str(hvd.rank())
 
 
+def init_sok():
+  try:
+    sok.init()
+    os.environ['ENABLE_SOK'] = '1'
```
Comment (poson) [inline easy_rec/python/utils/estimator_utils.py:1024]: ENABLE_SOK，HOROVOD_RANK
相关的文档也一次更新一下？

File: easy_rec/python/input/parquet_input_v2.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+  def __init__(self,
+               data_config,
+               feature_config,
+               input_path,
+               task_index=0,
+               task_num=1,
+               check_mode=False,
+               pipeline_config=None,
+               **kwargs):
+    super(ParquetInputV2,
```
Comment (poson) [inline easy_rec/python/input/parquet_input_v2.py:34]: ParquetInputV2 和ParquetInput的区别是什么？
在文档里面直接写一下？

File: easy_rec/python/inference/parquet_predictor.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+from easy_rec.python.input.parquet_input import ParquetInput
+from easy_rec.python.protos.dataset_pb2 import DatasetConfig
+from easy_rec.python.utils import config_util
+from easy_rec.python.utils import input_utils
+
+try:
+  from tensorflow.python.framework.load_library import load_op_library
+  import easy_rec
+  load_embed_lib_path = os.path.join(easy_rec.ops_dir, 'libload_embed.so')
+  load_embed_lib = load_op_library(load_embed_lib_path)
```
Comment (tiankongdeguiji) [inline easy_rec/python/inference/parquet_predictor.py:25]: maintain the source code of libload_embed and compile it in EasyRec to support different versions of tensorflow

File: easy_rec/python/input/load_parquet.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+      eid = sid + batch_size
+      data_dict = {}
+
+      # sid_stub = sid
+      # sid = sid + sub_batch_size * task_index
+      # eid = sid + sub_batch_size
+
+      if label_fields is not None and len(label_fields) > 0:
+        for k in label_fields:
+          data_dict[k] = np.array([x[0] for x in input_data[k][sid:eid]],
```
Comment (tiankongdeguiji) [inline easy_rec/python/input/load_parquet.py:139]: label and feature in parquet may not be a list, make it more general

File: easy_rec/python/input/parquet_input_v2.py [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+            tf.estimator.export.ServingInputReceiver instance
+      """
+      if mode in (tf.estimator.ModeKeys.TRAIN, tf.estimator.ModeKeys.EVAL,
+                  tf.estimator.ModeKeys.PREDICT):
+        # build dataset from self._config.input_path
+        self._mode = mode
+        dataset = self._build(mode, params)
+        return dataset
+      elif mode is None:  # serving_input_receiver_fn for export SavedModel
+        ragged_ids = array_ops.placeholder(tf.int64, [None], name='ragged_ids')
```
Comment (tiankongdeguiji) [inline easy_rec/python/input/parquet_input_v2.py:351]: serving inputs is not correct

File: easy_rec/python/input/parquet_input.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+# from tensorflow.python.ops import logging_ops
+# from tensorflow.python.ops import math_ops
+from tensorflow.python.platform import gfile
+
+from easy_rec.python.compat import queues
+from easy_rec.python.input import load_parquet
+from easy_rec.python.input.input import Input
+
+
+class ParquetInput(Input):
```
Comment (tiankongdeguiji) [inline easy_rec/python/input/parquet_input.py:23]: add docstring for choice between ParquetInput and ParquetInputV2

File: easy_rec/python/layers/input_layer.py [hunk 9]
```diff
@@ -243,7 +250,8 @@ def single_call_input_layer(self,
         features,
         group_columns,
         cols_to_output_tensors=cols_to_output_tensors,
-        feature_name_to_output_tensors=feature_name_to_output_tensors)
+        feature_name_to_output_tensors=feature_name_to_output_tensors,
+        is_training=self._is_training)
```
Comment (tiankongdeguiji) [inline easy_rec/python/layers/input_layer.py:254]: use is_training for?

File: easy_rec/python/main.py [hunk 10]
```diff
@@ -341,11 +340,18 @@ def _train_and_evaluate_impl(pipeline_config,
   # Currently only a single Eval Spec is allowed.
   train_spec = tf.estimator.TrainSpec(
       input_fn=train_input_fn, max_steps=train_steps)
-  # create eval spec
-  eval_spec = _create_eval_export_spec(
-      pipeline_config, eval_data, check_mode=check_mode)
-  from easy_rec.python.compat import estimator_train
-  estimator_train.train_and_evaluate(estimator, train_spec, eval_spec)
+  if estimator_utils.has_sok():
+    estimator.train(
```
Comment (tiankongdeguiji) [inline easy_rec/python/main.py:344]: why only call train when has_sok?

File: easy_rec/python/main.py [hunk 11]
```diff
@@ -793,6 +800,20 @@ def export(export_dir,
       checkpoint_path=ckpt_path,
       strip_default_attrs=True)
 
+  # part_embed_path = ckpt_path + '-embedding/'
```
Comment (tiankongdeguiji) [inline easy_rec/python/main.py:803]: remove

**Review by tiankongdeguiji — CHANGES_REQUESTED** [review]

File: easy_rec/python/layers/input_layer.py (hunk 9 above)
Comment (chengmengli06) [reply easy_rec/python/layers/input_layer.py:254]: for kv embedding_loookup, when is_training is false, return zero embedding vectors, and does not create a new entry in kv map

File: easy_rec/python/main.py (hunk 10 above)
Comment (chengmengli06) [reply easy_rec/python/main.py:344]: evaluation under multi-worker mode is not good now, currently only works under single worker mode.

File: easy_rec/python/utils/estimator_utils.py (hunk 3 above)
Comment (chengmengli06) [reply easy_rec/python/utils/estimator_utils.py:1024]: this are internal variables, only developers have to know about them.

File: easy_rec/python/utils/test_utils.py (hunk 2 above)
Comment (chengmengli06) [reply easy_rec/python/utils/test_utils.py:604]: means enable feature column sorting, by default it is not sorted, in order to support to warmup from old embedding checkpoints.

File: easy_rec/python/inference/parquet_predictor.py (hunk 5 above)
Comment (chengmengli06) [reply easy_rec/python/inference/parquet_predictor.py:25]: done

File: easy_rec/python/input/parquet_input_v2.py (hunk 4 above)
Comment (chengmengli06) [reply easy_rec/python/input/parquet_input_v2.py:33]: Explained in dataset.proto file, whether to pack sparse features and dense features.

File: easy_rec/python/input/parquet_input.py (hunk 8 above)
Comment (chengmengli06) [reply easy_rec/python/input/parquet_input.py:23]: add in dataset.proto file

File: easy_rec/python/utils/input_utils.py (hunk 1 above)
Comment (chengmengli06) [reply easy_rec/python/utils/input_utils.py]: done

File: easy_rec/python/main.py (hunk 11 above)
Comment (chengmengli06) [reply easy_rec/python/main.py]: done

**Review by poson — APPROVED** [review]

**Review by tiankongdeguiji — APPROVED** [review]
LGTM

**Pull request merged**